crypto/evp/evp_enc.c

```diff
@@ -452,6 +474,158 @@ static int evp_cipher_init_internal(EVP_CIPHER_CTX *ctx,
+ */
+static int evp_cipher_init_skey_internal(EVP_CIPHER_CTX *ctx,
+                                         const EVP_CIPHER *cipher,
+                                         const EVP_SKEY *skey,
+                                         const unsigned char *iv, size_t iv_len,
+                                         int enc, const OSSL_PARAM params[])
+{
+    int ret;
+
+    /*
+     * enc == 1 means we are encrypting.
+     * enc == 0 means we are decrypting.
+     * enc == -1 means, use the previously initialised value for encrypt/decrypt
+     */
+    if (enc == -1) {
+        enc = ctx->encrypt;
+    } else {
+        if (enc)
+            enc = 1;
+        ctx->encrypt = enc;
+    }
+
+    if (cipher == NULL && ctx->cipher == NULL) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_NO_CIPHER_SET);
+        return 0;
+    }
+
+    /*
+     * If there are engines involved then we throw an error
+     */
+    if (ctx->engine != NULL
+            || (cipher != NULL && cipher->origin == EVP_ORIG_METH)
+            || (cipher == NULL && ctx->cipher != NULL
+                && ctx->cipher->origin == EVP_ORIG_METH)) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_INITIALIZATION_ERROR);
+        return 0;
+    }
+    /*
+     * Ensure a context left lying around from last time is cleared
+     * (legacy code)
+     */
+    if (cipher != NULL && ctx->cipher != NULL) {
+        if (ctx->cipher->cleanup != NULL && !ctx->cipher->cleanup(ctx))
+            return 0;
+        OPENSSL_clear_free(ctx->cipher_data, ctx->cipher->ctx_size);
+        ctx->cipher_data = NULL;
+    }
+
+    /* Ensure a context left lying around from last time is cleared */
+    if (cipher != NULL && ctx->cipher != NULL) {
+        unsigned long flags = ctx->flags;
+
+        EVP_CIPHER_CTX_reset(ctx);
+        /* Restore encrypt and flags */
+        ctx->encrypt = enc;
+        ctx->flags = flags;
+    }
+
+    if (cipher == NULL)
+        cipher = ctx->cipher;
+
+    if (cipher->prov == NULL) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_INITIALIZATION_ERROR);
+        return 0;
+    }
+
+    if (cipher != ctx->fetched_cipher) {
+        if (!EVP_CIPHER_up_ref((EVP_CIPHER *)cipher)) {
+            ERR_raise(ERR_LIB_EVP, EVP_R_INITIALIZATION_ERROR);
+            return 0;
+        }
+        EVP_CIPHER_free(ctx->fetched_cipher);
+        /* Coverity false positive, the reference counting is confusing it */
+        /* coverity[use_after_free] */
+        ctx->fetched_cipher = (EVP_CIPHER *)cipher;
+    }
+    ctx->cipher = cipher;
+    if (ctx->algctx == NULL) {
+        ctx->algctx = ctx->cipher->newctx(ossl_provider_ctx(cipher->prov));
+        if (ctx->algctx == NULL) {
+            ERR_raise(ERR_LIB_EVP, EVP_R_INITIALIZATION_ERROR);
+            return 0;
+        }
+    }
+
+    if (skey != NULL && skey->skeymgmt != NULL && ctx->cipher->prov != skey->skeymgmt->prov) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_INITIALIZATION_ERROR);
+        return 0;
+    }
```

Comment:
@levitte do you think that this check is not enough? I can add extra check similar to https://github.com/openssl/openssl/blob/d69c014608acdfa37839d49412e6d6974ac539a0/crypto/evp/signature.c#L558-L569 there if you insist

Reply:
Assuming that a provider supports more than one symmetric key type, not checking enables a caller to use an inappropriate key type for the cipher.

For example, if a provider includes the key size in the key type name (`AES-128`, `AES-192`), you could potentially try to use an AES-192 key with AES-128-CBC.

Another potential example, say a provider that supports two distinct algorithms, such as SEED and CAMELLIA, what would stop a caller from trying to use a SEED key with CAMELLIA?

So yeah, I think that added check is necessary.

Reply:
The only reason we didn't have this check for keypair schemes was that we composed them by combining the key type and the digest, so we didn't have a name mismatch.

The keypair check became crucial as soon as we started supported composite scheme names (the EVP_PKEY_sign effort, if you remember), like RSA-SHA256.

Reply:
@levitte the provider should check the key type at cipher initialization and fail.
I do not know that making arbitrary checks in libcrypto is necessarily a good idea or useful. The provider knows what key types can be used by the requested mechanisms.

For example in PKCS#11 some mechanism can both have a specific key type but also accept using the generic secret key type for some operations (as long as they have a proper length, which can be a range and not discrete values in many cases).

Reply:
I understand I can return a long list of strings, but at last for this case, it seems a superfluous check, as the provider will have to re-check these things anyway.

Is there a case where you would need libcrypto to make this extra check?
Perhaps to avoid a full import/export cycle when trying cross-provider operations?

Reply:
> Perhaps to avoid a full import/export cycle when trying cross-provider operations?

There's no avoiding that, really, 'cause one provider cannot be assumed to know the internal storage of another provider.  So no.

So uhm, I don't have a particular user case in mind for the moment...  and perhaps your way is what should have been done for keypairs as well (unless I forget something).

Reply:
So if you don't insist, and resolve the issue, I will drop back my last commit :)

Reply:
Yes, please drop any key type checking on libcrypto side.

Reply:
With pleasure :)
Done, thanks